crates/viewer/re_redap_browser/src/servers.rs

```diff
@@ -104,158 +212,57 @@ impl RedapServers {
-                    ui,
-                    egui::Id::new(origin).with("server_item"),
-                    true,
-                    content,
-                    |ui| {
-                        self.catalog_list_ui(ui, origin, catalog);
-                    },
-                );
-        }
-    }
+        //TODO(ab): we should display something even if no catalog is currently selected.
 
-    fn catalog_list_ui(&self, ui: &mut egui::Ui, origin: &re_uri::Origin, catalog: &Catalog) {
-        if catalog.is_empty() {
-            ui.list_item_flat_noninteractive(list_item::LabelContent::new("(empty)").italics(true));
-        } else {
-            for (index, collection) in catalog.collections.iter().enumerate() {
-                let is_selected =
-                    if let Some(selected_collection) = self.selected_collection.as_ref() {
-                        selected_collection.server_origin == *origin
-                            && selected_collection.collection_index == index
-                    } else {
-                        false
-                    };
-
-                let content = list_item::LabelContent::new(&collection.name);
-                let response = ui.list_item().selected(is_selected).show_flat(ui, content);
+        if let Some(selected_collection) = self.selected_collection.as_ref() {
+            for server in self.servers.values() {
+                let collection = server.find_collection(*selected_collection);
+
+                if let Some(collection) = collection {
+                    self.with_ctx(|ctx| {
+                        super::collection_ui::collection_ui(
+                            viewer_ctx,
+                            ctx,
+                            ui,
+                            &server.origin,
+                            collection,
+                        );
+                    });
 
-                if response.clicked() {
-                    self.command_queue
-                        .lock()
-                        .push(Command::SelectCollection(CollectionHandle {
-                            server_origin: origin.clone(),
-                            collection_index: index,
-                        }));
+                    return;
                 }
             }
         }
-
-        // deselect when clicking in empty space
-        let empty_space_response = ui.allocate_response(ui.available_size(), egui::Sense::click());
-
-        // clear selection upon clicking the empty space
-        if empty_space_response.clicked() {
-            self.command_queue.lock().push(Command::DeselectCollection);
-        }
     }
 
-    pub fn ui(&self, ctx: &ViewerContext<'_>, ui: &mut egui::Ui) {
-        //TODO(ab): we should display something even if no catalog is currently selected.
+    fn add_server_modal_ui(&mut self, ui: &egui::Ui) {
+        //TODO(ab): borrow checker doesn't let me use `with_ctx()` here, I should find a better way
```

Comment:
ah that pattern is tricky, don't know either.  Oooooonnneeee day the Polonius borrow checker will solve this, any decade now

Reply:
in fairness, pulling that off would be pretty great!